Classes/common/Encryption/Keychain/CDTEncryptionKeychainStorage.m

```diff
@@ -0,0 +1,198 @@
+
+    NSData *archivedData = [NSKeyedArchiver archivedDataWithRootObject:data];
+    NSMutableDictionary *dataStoreDict =
+        [CDTEncryptionKeychainStorage genericPwStoreDictWithService:self.service
+                                                            account:self.account
+                                                               data:archivedData];
+
+    OSStatus err = SecItemAdd((__bridge CFDictionaryRef)dataStoreDict, nil);
+    if (err == noErr) {
+        worked = YES;
+    } else if (err == errSecDuplicateItem) {
+        CDTLogWarn(CDTDATASTORE_LOG_CONTEXT, @"Doc already exists in keychain");
+        worked = NO;
+    } else {
+        CDTLogWarn(CDTDATASTORE_LOG_CONTEXT,
+                   @"Unable to store Doc in keychain, SecItemAdd returned: %d", err);
+        worked = NO;
+    }
+
+    return worked;
+}
+
+- (BOOL)clearEncryptionKeyData
+{
+    BOOL worked = NO;
+
+    NSMutableDictionary *dict =
+        [CDTEncryptionKeychainStorage genericPwLookupDictWithService:self.service
+                                                             account:self.account];
+    [dict removeObjectForKey:(__bridge id)(kSecMatchLimit)];
+    [dict removeObjectForKey:(__bridge id)(kSecReturnAttributes)];
+    [dict removeObjectForKey:(__bridge id)(kSecReturnData)];
+
+    OSStatus err = SecItemDelete((__bridge CFDictionaryRef)dict);
+
+    if (err == noErr || err == errSecItemNotFound) {
+        worked = YES;
+    } else {
+        CDTLogWarn(CDTDATASTORE_LOG_CONTEXT,
+                   @"Error getting DPK doc from keychain, SecItemDelete returned: %d", err);
+    }
+
+    return worked;
+}
+
+- (BOOL)areThereEncryptionKeyData
+{
+    BOOL result = NO;
+
+    NSData *data = nil;
+    NSMutableDictionary *lookupDict =
+        [CDTEncryptionKeychainStorage genericPwLookupDictWithService:self.service
+                                                             account:self.account];
+
+    OSStatus err = SecItemCopyMatching((__bridge CFDictionaryRef)lookupDict, (void *)&data);
+    if (err == noErr) {
+        result = ((data != nil) && (data.length > 0));
+
+        if (!result) {
+            CDTLogWarn(CDTDATASTORE_LOG_CONTEXT, @"Found a match in keychain, but it was empty");
+        }
+    } else if (err == errSecItemNotFound) {
+        CDTLogWarn(CDTDATASTORE_LOG_CONTEXT, @"DPK doc not found in keychain");
+    } else {
+        CDTLogWarn(CDTDATASTORE_LOG_CONTEXT,
+                   @"Error getting DPK doc from keychain, SecItemCopyMatching returned: %d", err);
+    }
+
+    return result;
+}
+
+#pragma mark - Private class methods
++ (NSMutableDictionary *)genericPwLookupDictWithService:(NSString *)service
+                                                account:(NSString *)account
+{
+    NSMutableDictionary *genericPasswordQuery = [[NSMutableDictionary alloc] init];
+    
+    [genericPasswordQuery setObject:(__bridge id)kSecClassGenericPassword
+                             forKey:(__bridge id)kSecClass];
```

Comment:
Also, `kSecClassKey ` is intended to store a cryptographic key but we are not doing that, we are storing the means to decipher a cryptographic key ... I hope this helps to understand the reason for this choice.